src/main/java/fi/aalto/cs/intellij/common/Course.java

```diff
@@ -0,0 +1,219 @@
+   * plugins and the values are the names corresponding to the ids.
+   *
+   * @return A map with the required plugins of the course.
+   */
+  @NotNull
+  public Map<String, String> getRequiredPlugins() {
+    return requiredPlugins;
+  }
+
+  @NotNull
+  private static JSONObject getCourseJsonObject(@NotNull String path)
+      throws FileNotFoundException, MalformedCourseConfigurationFileException {
+    FileReader file = new FileReader(path);
+    JSONTokener tokenizer = new JSONTokener(file);
+    try {
+      return new JSONObject(tokenizer);
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Course configuration file should consist of a valid JSON object", ex);
+    }
+  }
+
+  @NotNull
+  private static String getCourseName(@NotNull JSONObject jsonObject, @NotNull String path)
+      throws MalformedCourseConfigurationFileException {
+    try {
+      return jsonObject.getString("name");
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Missing or malformed \"name\" key", ex);
+    }
+  }
+
+  @NotNull
+  private static List<Module> getCourseModules(JSONObject jsonObject, String path)
+      throws MalformedCourseConfigurationFileException {
+    JSONArray modulesJsonArray;
+    try {
+      modulesJsonArray = jsonObject.getJSONArray("modules");
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Missing or malformed \"modules\" key", ex);
+    }
+
+    List<Module> modules = new ArrayList<>();
+    // Indexing loop used to simplify checking that each entry is a JSON object.
+    for (int i = 0; i < modulesJsonArray.length(); ++i) {
+      JSONObject moduleObject = modulesJsonArray.optJSONObject(i);
+      if (moduleObject == null) {
+        throw new MalformedCourseConfigurationFileException(path,
+            "\"modules\" value should be an array of objects", null);
+      }
+      modules.add(getIndividualModule(moduleObject, path));
+    }
+    return modules;
+  }
+
+  @NotNull
+  private static Module getIndividualModule(@NotNull JSONObject moduleObject, @NotNull String path)
+      throws MalformedCourseConfigurationFileException {
+    try {
+      String moduleName = moduleObject.getString("name");
+      URL moduleUrl = new URL(moduleObject.getString("url"));
+      return new Module(moduleName, moduleUrl);
+    } catch (JSONException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Module objects should contain \"name\" and \"url\" keys with string values", ex);
+    } catch (MalformedURLException ex) {
+      throw new MalformedCourseConfigurationFileException(path,
+          "Malformed URL in module object", ex);
+    }
+  }
+
+  @NotNull
+  private static Map<String, String> getCourseRequiredPlugins(@NotNull JSONObject jsonObject,
```

Comment:
this method might still be quite challenging to test, I suggest considering splitting it even more